src/openag_brain/software_modules/recipe_handler.py

```diff
@@ -12,248 +12,276 @@
+        start_time = start_time or init_time
+        if start_time > init_time:
+            raise ValueError("Recipes cannot be scheduled for the future")
+        self.start_time = start_time
+        self.id = recipe["_id"]
+        self.operations = recipe["operations"]
+        self.timeout = timeout
 
-    def __getitem__(self, topic):
-        if not topic in self.publishers:
-            self.publishers[topic] = rospy.Publisher(
-                "desired/"+topic, Float64, queue_size=10
-            )
-        return self.publishers[topic]
+    def __iter__(self):
+        """
+        Create a blocking recipe generator for simple recipes.
+        Yields a series of setpoints for current time.
+        """
+        # Create a state object to accrue recipe setpoint values.
+        state = {}
+        # If start time was now (or in the very recent past), yield
+        # a recipe start setpoint.
+        if time.time() - self.start_time < 1:
+            yield (self.start_time, RECIPE_START.name, self.id)
+        for t, variable, value in self.operations:
+            # While we wait for time to catch up to timestamp, yield the
+            # previous state once every second.
+            while t > time.time() - self.start_time:
+                # Make sure the variable names in this inner loop are different
+                # from the ones above. Python scopes loop variables to the
+                # function level (not the loop level). Ran afoul of scoping
+                # bug here in the past.
+                # http://eli.thegreenplace.net/2015/the-scope-of-index-variables-in-pythons-for-loops/
+                # 2017-01-09 @gordonbrander
+                for state_variable, state_value in state.iteritems():
+                    yield (time.time(), state_variable, state_value)
+                rospy.sleep(self.timeout)
+            # Ok, setpoint has reached the present. Assign it to state.
+            # Then loop until we hit a future state, at which point, we
+            # start yielding the present state again.
+            state[variable] = value
+        # We're done! Yield any final state changes we picked up
+        # on the last iteration, then yield a RECIPE_END setpoint.
+        for state_variable, state_value in state.iteritems():
+            yield (time.time(), variable, value)
+        yield (time.time(), RECIPE_END.name, self.id)
+
+class RecipeRunningError(Exception):
+    """Thrown when trying to set a recipe, but recipe is already running."""
+    pass
 
-class RecipeHandler(object):
-    # Dictionary that maps recipe formats to python classes
-    recipe_class_map = {
-        "simple": SimpleRecipe
-    }
+class RecipeIdleError(Exception):
+    """Thrown when trying to clear a recipe, but recipe is already clear."""
+    pass
 
-    def __init__(self, server):
+class RecipeHandler:
+    def __init__(self, server, environment):
+        # We create a lock to ensure threadsafety, since service handlers are
```

Comment:
I wasn't sure if ROS ran services in processes (independent of GIL) or threads (dependent on GIL), so this might just be paranoia. I would like to keep this for correctness anyway, if that's fine with you.